### PR TITLE
Fix test command teardown order

### PR DESCRIPTION
**Pull request by Evi1Pumpkin**
## The problem
Today, the `tofu test` command creates resources and then tears them down. It first destroys the resources in the state of the "main" run, and then goes over the the states of any additional modules and destroys them in reverse to their run order.

So if the run blocks order is:
"Module A" -> "Main configuration" -> "Module B"
The cleanup order will be:
"Main configuration" -> "Module B" -> "Module A"
If "Module B" has a data resource referencing "Main configuration", we'll face an error on cleanup (as discussed [here](https://github.com/opentofu/opentofu/issues/1013)).

## The solution
Change the cleanup order. Don't destroy the "main" state first, instead destroy based on running order:
 "Module B"  -> "Main configuration" -> "Module A".

Resolves https://github.com/opentofu/opentofu/issues/1013

## Target Release

1.7.0

### CODE REVIEW

File: internal/command/test_test.go [hunk 1]
```diff
@@ -780,58 +780,98 @@ can remove the provider configuration again.
 	}
 }
 
-func TestTest_NestedSetupModules(t *testing.T) {
-	td := t.TempDir()
-	testCopyDir(t, testFixturePath(path.Join("test", "with_nested_setup_modules")), td)
-	defer testChdir(t, td)()
+func TestTest_Modules(t *testing.T) {
```
Comment (Evi1Pumpkin) [inline internal/command/test_test.go:783]: I merged a few similar tests together - `TestTest_OnlyExternalModules`, `TestTest_NestedSetupModules` and the new `with_verify_module`

Comment (cube2222) [reply internal/command/test_test.go:783]: Do I understand correctly that this new test is a regression test that fails before the change?

**Review by cube2222 — APPROVED** [review]
LGTM!

File: internal/command/test_test.go (hunk 1 above)
Comment (Evi1Pumpkin) [reply internal/command/test_test.go:783]: Indeed 🙃

**Review by cam72cam — APPROVED** [review]

File: internal/command/testdata/test/with_verify_module/main.tftest.hcl [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+
+run "test" {
+}
+
+run "verify" {
+  module {
+    source = "./verify"
+  }
+
+ assert {
```
Comment (janosdebugs) [inline internal/command/testdata/test/with_verify_module/main.tftest.hcl:14]: The formatting seems to be slightly off here?

**Review by janosdebugs — APPROVED** [review]

File: internal/command/testdata/test/with_verify_module/main.tftest.hcl (hunk 2 above)
Comment (Evi1Pumpkin) [reply internal/command/testdata/test/with_verify_module/main.tftest.hcl]: You are sharp 👁️ 
Fixed https://github.com/opentofu/opentofu/pull/1043/commits/d3d39f2b3b743ffadfca90c8172d2a536f8363dc

**Pull request merged**